## ANA-522
## Mod02 week04: pandas 

### Explore Pokemon Dataset
[The complete Pokemon Dataset](https://www.kaggle.com/rounakbanik/pokemon) is available on [Kaggle.com](http://kaggle.com)  

I also attached the dataset on Blackboard.

This dataset contains information on all 802 Pokemon from all Seven Generations of Pokemon. The information contained in this dataset include Base Stats, Performance against Other Types, Height, Weight, Classification, Egg Steps, Experience Points, Abilities, etc. The information was scraped from http://serebii.net/

We are going to use pandas and some NumPy functions to explore the dataset and extract information by manipulating data points from featured attributes. 

<br>

## Step 1: Load the pandas and numpy libraries.

- `import pandas as pd`: This line imports the pandas library and gives it the alias "pd". This means that whenever we want to use a function from the pandas library, we can prefix it with "pd." instead of typing out "pandas."



- `import numpy as np`: Similarly, this line imports the numpy library and gives it the alias "np". 

In [1]:
import pandas as pd
import numpy as np

## Step 2: Load the dataset.

We'll use the `pd.read_csv()` function from the pandas library to load the CSV file into a DataFrame. 

If you're running the notebook locally on your computer, you need to have a local copy of pokemon.csv in the same folder as your notebook file. You can then use the following statement to load the data:

By specifying `sep=','`, we're telling pandas to expect each value in the file to be separated by a comma. If the values in the file were separated by a different character, like a semicolon ;, we would use `sep=';'` instead.

In [3]:
# Loading the dataset into a DataFrame named "pokemon"

pokemon = pd.read_csv('/Users/nkohei/Workspace/McDaniel-Repository/522/week4/pokemon.csv', sep = ',')

Or, if your data file is in a different folder or path, you can specify the full file path to read it. For example:

In [ ]:
# For Windows:

# pokemon = pd.read_csv('C:/path_to_your_file/pokemon.csv', sep=',')

In [ ]:
# For Mac or Linux:

# pokemon = pd.read_csv('/path_to_your_file/pokemon.csv', sep=',')

<br>

### Displaying the first 5 rows of the dataset for a quick overview Using `.head()` function

In [5]:
pokemon.head() 

abilities  against_bug  against_dark  against_dragon  \
0  ['Overgrow', 'Chlorophyll']          1.0           1.0             1.0   
1  ['Overgrow', 'Chlorophyll']          1.0           1.0             1.0   
2  ['Overgrow', 'Chlorophyll']          1.0           1.0             1.0   
3     ['Blaze', 'Solar Power']          0.5           1.0             1.0   
4     ['Blaze', 'Solar Power']          0.5           1.0             1.0   

   against_electric  against_fairy  against_fight  against_fire  \
0               0.5            0.5            0.5           2.0   
1               0.5            0.5            0.5           2.0   
2               0.5            0.5            0.5           2.0   
3               1.0            0.5            1.0           0.5   
4               1.0            0.5            1.0           0.5   

   against_flying  against_ghost  ...  percentage_male  pokedex_number  \
0             2.0            1.0  ...             88.1               1   
1             2.0            1.0  ...             88.1               2   
2             2.0            1.0  ...             88.1               3   
3             1.0            1.0  ...             88.1               4   
4             1.0            1.0  ...             88.1               5   

   sp_attack  sp_defense  speed  type1   type2  weight_kg  generation  \
0         65          65     45  grass  poison        6.9           1   
1         80          80     60  grass  poison       13.0           1   
2        122         120     80  grass  poison      100.0           1   
3         60          50     65   fire     NaN        8.5           1   
4         80          65     80   fire     NaN       19.0           1   

   is_legendary  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 41 columns]

### Display the first 10 records in the Pokemon Dataset.  

In [6]:
pokemon.head(10) 

abilities  against_bug  against_dark  against_dragon  \
0  ['Overgrow', 'Chlorophyll']         1.00           1.0             1.0   
1  ['Overgrow', 'Chlorophyll']         1.00           1.0             1.0   
2  ['Overgrow', 'Chlorophyll']         1.00           1.0             1.0   
3     ['Blaze', 'Solar Power']         0.50           1.0             1.0   
4     ['Blaze', 'Solar Power']         0.50           1.0             1.0   
5     ['Blaze', 'Solar Power']         0.25           1.0             1.0   
6     ['Torrent', 'Rain Dish']         1.00           1.0             1.0   
7     ['Torrent', 'Rain Dish']         1.00           1.0             1.0   
8     ['Torrent', 'Rain Dish']         1.00           1.0             1.0   
9  ['Shield Dust', 'Run Away']         1.00           1.0             1.0   

   against_electric  against_fairy  against_fight  against_fire  \
0               0.5            0.5            0.5           2.0   
1               0.5            0.5            0.5           2.0   
2               0.5            0.5            0.5           2.0   
3               1.0            0.5            1.0           0.5   
4               1.0            0.5            1.0           0.5   
5               2.0            0.5            0.5           0.5   
6               2.0            1.0            1.0           0.5   
7               2.0            1.0            1.0           0.5   
8               2.0            1.0            1.0           0.5   
9               1.0            1.0            0.5           2.0   

   against_flying  against_ghost  ...  percentage_male  pokedex_number  \
0             2.0            1.0  ...             88.1               1   
1             2.0            1.0  ...             88.1               2   
2             2.0            1.0  ...             88.1               3   
3             1.0            1.0  ...             88.1               4   
4             1.0            1.0  ...             88.1               5   
5             1.0            1.0  ...             88.1               6   
6             1.0            1.0  ...             88.1               7   
7             1.0            1.0  ...             88.1               8   
8             1.0            1.0  ...             88.1               9   
9             2.0            1.0  ...             50.0              10   

   sp_attack  sp_defense  speed  type1   type2  weight_kg  generation  \
0         65          65     45  grass  poison        6.9           1   
1         80          80     60  grass  poison       13.0           1   
2        122         120     80  grass  poison      100.0           1   
3         60          50     65   fire     NaN        8.5           1   
4         80          65     80   fire     NaN       19.0           1   
5        159         115    100   fire  flying       90.5           1   
6         50          64     43  water     NaN        9.0           1   
7         65          80     58  water     NaN       22.5           1   
8        135         115     78  water     NaN       85.5           1   
9         20          20     45    bug     NaN        2.9           1   

   is_legendary  
0             0  
1             0  
2             0  
3             0  
4             0  
5             0  
6             0  
7             0  
8             0  
9             0  

[10 rows x 41 columns]

### Or you would simply use the name `pokemon` in a code cell to display the dataset in Jupyter Notebook.

In [7]:
pokemon

abilities  against_bug  against_dark  against_dragon  \
0    ['Overgrow', 'Chlorophyll']         1.00           1.0             1.0   
1    ['Overgrow', 'Chlorophyll']         1.00           1.0             1.0   
2    ['Overgrow', 'Chlorophyll']         1.00           1.0             1.0   
3       ['Blaze', 'Solar Power']         0.50           1.0             1.0   
4       ['Blaze', 'Solar Power']         0.50           1.0             1.0   
..                           ...          ...           ...             ...   
796              ['Beast Boost']         0.25           1.0             0.5   
797              ['Beast Boost']         1.00           1.0             0.5   
798              ['Beast Boost']         2.00           0.5             2.0   
799              ['Prism Armor']         2.00           2.0             1.0   
800               ['Soul-Heart']         0.25           0.5             0.0   

     against_electric  against_fairy  against_fight  against_fire  \
0                 0.5            0.5            0.5           2.0   
1                 0.5            0.5            0.5           2.0   
2                 0.5            0.5            0.5           2.0   
3                 1.0            0.5            1.0           0.5   
4                 1.0            0.5            1.0           0.5   
..                ...            ...            ...           ...   
796               2.0            0.5            1.0           2.0   
797               0.5            0.5            2.0           4.0   
798               0.5            4.0            2.0           0.5   
799               1.0            1.0            0.5           1.0   
800               1.0            0.5            1.0           2.0   

     against_flying  against_ghost  ...  percentage_male  pokedex_number  \
0               2.0            1.0  ...             88.1               1   
1               2.0            1.0  ...             88.1               2   
2               2.0            1.0  ...             88.1               3   
3               1.0            1.0  ...             88.1               4   
4               1.0            1.0  ...             88.1               5   
..              ...            ...  ...              ...             ...   
796             0.5            1.0  ...              NaN             797   
797             1.0            1.0  ...              NaN             798   
798             1.0            0.5  ...              NaN             799   
799             1.0            2.0  ...              NaN             800   
800             0.5            1.0  ...              NaN             801   

     sp_attack  sp_defense  speed    type1   type2  weight_kg  generation  \
0           65          65     45    grass  poison        6.9           1   
1           80          80     60    grass  poison       13.0           1   
2          122         120     80    grass  poison      100.0           1   
3           60          50     65     fire     NaN        8.5           1   
4           80          65     80     fire     NaN       19.0           1   
..         ...         ...    ...      ...     ...        ...         ...   
796        107         101     61    steel  flying      999.9           7   
797         59          31    109    grass   steel        0.1           7   
798         97          53     43     dark  dragon      888.0           7   
799        127          89     79  psychic     NaN      230.0           7   
800        130         115     65    steel   fairy       80.5           7   

     is_legendary  
0               0  
1               0  
2               0  
3               0  
4               0  
..            ...  
796             1  
797             1  
798             1  
799             1  
800             1  

[801 rows x 41 columns]

<br>

## Step 3: Exploring the Dataset.

### Check the shape of the dataset using `.shape` to see how many rows and columns it has.

In [8]:
# Checking the shape of the dataset
pokemon.shape

(801, 41)

The dataset contains 801 rows and 41 columns.

<br>

### Use `.info()` to get a summary of the dataset, including the number of non-null values in each column and data types.

In [9]:
pokemon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 801 entries, 0 to 800
Data columns (total 41 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   abilities          801 non-null    object 
 1   against_bug        801 non-null    float64
 2   against_dark       801 non-null    float64
 3   against_dragon     801 non-null    float64
 4   against_electric   801 non-null    float64
 5   against_fairy      801 non-null    float64
 6   against_fight      801 non-null    float64
 7   against_fire       801 non-null    float64
 8   against_flying     801 non-null    float64
 9   against_ghost      801 non-null    float64
 10  against_grass      801 non-null    float64
 11  against_ground     801 non-null    float64
 12  against_ice        801 non-null    float64
 13  against_normal     801 non-null    float64
 14  against_poison     801 non-null    float64
 15  against_psychic    801 non-null    float64
 16  against_rock       801 non

Here's a summary of what we see from `pokemon.info()`:

- There are a total of 41 columns in the dataset.
- The columns have different data types: float64 (21 columns), int64 (13 columns), and object (7 columns).
- Not all columns have 801 non-null values, indicating that some columns have missing values. For instance:
    - `height_m` has 781 non-null values, which means there are 20 missing values in this column.
    - `percentage_male` has 703 non-null values.
    - `type2` has 417 non-null values, indicating that not all Pokémon have a secondary type.

### Or we can also use `.dtypes` to obtain the data type of each column in the DataFrame.

In [10]:
pokemon.dtypes

abilities             object
against_bug          float64
against_dark         float64
against_dragon       float64
against_electric     float64
against_fairy        float64
against_fight        float64
against_fire         float64
against_flying       float64
against_ghost        float64
against_grass        float64
against_ground       float64
against_ice          float64
against_normal       float64
against_poison       float64
against_psychic      float64
against_rock         float64
against_steel        float64
against_water        float64
attack                 int64
base_egg_steps         int64
base_happiness         int64
base_total             int64
capture_rate          object
classfication         object
defense                int64
experience_growth      int64
height_m             float64
hp                     int64
japanese_name         object
name                  object
percentage_male      float64
pokedex_number         int64
sp_attack              int64
sp_defense    

<br>

### Use `.describe()` to get statistical summaries.

In [11]:
pokemon.describe()

against_bug  against_dark  against_dragon  against_electric  \
count   801.000000    801.000000      801.000000        801.000000   
mean      0.996255      1.057116        0.968789          1.073970   
std       0.597248      0.438142        0.353058          0.654962   
min       0.250000      0.250000        0.000000          0.000000   
25%       0.500000      1.000000        1.000000          0.500000   
50%       1.000000      1.000000        1.000000          1.000000   
75%       1.000000      1.000000        1.000000          1.000000   
max       4.000000      4.000000        2.000000          4.000000   

       against_fairy  against_fight  against_fire  against_flying  \
count     801.000000     801.000000    801.000000      801.000000   
mean        1.068976       1.065543      1.135456        1.192884   
std         0.522167       0.717251      0.691853        0.604488   
min         0.250000       0.000000      0.250000        0.250000   
25%         1.000000       0.500000      0.500000        1.000000   
50%         1.000000       1.000000      1.000000        1.000000   
75%         1.000000       1.000000      2.000000        1.000000   
max         4.000000       4.000000      4.000000        4.000000   

       against_ghost  against_grass  ...    height_m          hp  \
count     801.000000     801.000000  ...  781.000000  801.000000   
mean        0.985019       1.034020  ...    1.163892   68.958801   
std         0.558256       0.788896  ...    1.080326   26.576015   
min         0.000000       0.250000  ...    0.100000    1.000000   
25%         1.000000       0.500000  ...    0.600000   50.000000   
50%         1.000000       1.000000  ...    1.000000   65.000000   
75%         1.000000       1.000000  ...    1.500000   80.000000   
max         4.000000       4.000000  ...   14.500000  255.000000   

       percentage_male  pokedex_number   sp_attack  sp_defense       speed  \
count       703.000000      801.000000  801.000000  801.000000  801.000000   
mean         55.155761      401.000000   71.305868   70.911361   66.334582   
std          20.261623      231.373075   32.353826   27.942501   28.907662   
min           0.000000        1.000000   10.000000   20.000000    5.000000   
25%          50.000000      201.000000   45.000000   50.000000   45.000000   
50%          50.000000      401.000000   65.000000   66.000000   65.000000   
75%          50.000000      601.000000   91.000000   90.000000   85.000000   
max         100.000000      801.000000  194.000000  230.000000  180.000000   

        weight_kg  generation  is_legendary  
count  781.000000  801.000000    801.000000  
mean    61.378105    3.690387      0.087391  
std    109.354766    1.930420      0.282583  
min      0.100000    1.000000      0.000000  
25%      9.000000    2.000000      0.000000  
50%     27.300000    4.000000      0.000000  
75%     64.800000    5.000000      0.000000  
max    999.900000    7.000000      1.000000  

[8 rows x 34 columns]

<br>

### Use the `.columns` attribute to obtain an overview of the column names (or labels) present in the pokemon DataFrame

In [12]:
pokemon.columns

Index(['abilities', 'against_bug', 'against_dark', 'against_dragon',
       'against_electric', 'against_fairy', 'against_fight', 'against_fire',
       'against_flying', 'against_ghost', 'against_grass', 'against_ground',
       'against_ice', 'against_normal', 'against_poison', 'against_psychic',
       'against_rock', 'against_steel', 'against_water', 'attack',
       'base_egg_steps', 'base_happiness', 'base_total', 'capture_rate',
       'classfication', 'defense', 'experience_growth', 'height_m', 'hp',
       'japanese_name', 'name', 'percentage_male', 'pokedex_number',
       'sp_attack', 'sp_defense', 'speed', 'type1', 'type2', 'weight_kg',
       'generation', 'is_legendary'],
      dtype='object')

### Use the `.values` attribute to return the data in the pokemon DataFrame as a 2-dimensional numpy array, where rows represent records and columns represent features

In [13]:
pokemon.values

array([["['Overgrow', 'Chlorophyll']", 1.0, 1.0, ..., 6.9, 1, 0],
       ["['Overgrow', 'Chlorophyll']", 1.0, 1.0, ..., 13.0, 1, 0],
       ["['Overgrow', 'Chlorophyll']", 1.0, 1.0, ..., 100.0, 1, 0],
       ...,
       ["['Beast Boost']", 2.0, 0.5, ..., 888.0, 7, 1],
       ["['Prism Armor']", 2.0, 2.0, ..., 230.0, 7, 1],
       ["['Soul-Heart']", 0.25, 0.5, ..., 80.5, 7, 1]], dtype=object)

<br>

### Extract the "name" column from the pokemon DataFrame.

In [14]:
name_column = pokemon["name"]
name_column

0       Bulbasaur
1         Ivysaur
2        Venusaur
3      Charmander
4      Charmeleon
          ...    
796    Celesteela
797       Kartana
798      Guzzlord
799      Necrozma
800      Magearna
Name: name, Length: 801, dtype: object

<br>

### Access the data of the Pokémon named "Pikachu"

#### We can use `.loc` to filter the rows based on the "name" column and get all columns related to "Pikachu":

In [15]:
pokemon[pokemon["name"]=='Pikachu']

abilities  against_bug  against_dark  against_dragon  \
24  ['Static', 'Lightningrod']          1.0           1.0             1.0   

    against_electric  against_fairy  against_fight  against_fire  \
24               0.5            1.0            1.0           1.0   

    against_flying  against_ghost  ...  percentage_male  pokedex_number  \
24             0.5            1.0  ...             50.0              25   

    sp_attack  sp_defense  speed     type1  type2  weight_kg  generation  \
24         50          50     90  electric    NaN        6.0           1   

    is_legendary  
24             0  

[1 rows x 41 columns]

In [20]:
pikachu_data = pokemon.loc[pokemon['name'] == 'Pikachu']
pikachu_data

abilities  against_bug  against_dark  against_dragon  \
24  ['Static', 'Lightningrod']          1.0           1.0             1.0   

    against_electric  against_fairy  against_fight  against_fire  \
24               0.5            1.0            1.0           1.0   

    against_flying  against_ghost  ...  percentage_male  pokedex_number  \
24             0.5            1.0  ...             50.0              25   

    sp_attack  sp_defense  speed     type1  type2  weight_kg  generation  \
24         50          50     90  electric    NaN        6.0           1   

    is_legendary  
24             0  

[1 rows x 41 columns]

<br>

### Display all records in the Pokemon Dataset with name, type1, and type2 attributes only.

#### [['name', 'type1', 'type2']]: This is a list of column names that we want to extract from the DataFrame.

In [16]:
pokemon[['name','type1','type2']]

name    type1   type2
0     Bulbasaur    grass  poison
1       Ivysaur    grass  poison
2      Venusaur    grass  poison
3    Charmander     fire     NaN
4    Charmeleon     fire     NaN
..          ...      ...     ...
796  Celesteela    steel  flying
797     Kartana    grass   steel
798    Guzzlord     dark  dragon
799    Necrozma  psychic     NaN
800    Magearna    steel   fairy

[801 rows x 3 columns]

In [17]:
#Or we can also use:
pokemon.loc[:,['name','type1','type2']]

name    type1   type2
0     Bulbasaur    grass  poison
1       Ivysaur    grass  poison
2      Venusaur    grass  poison
3    Charmander     fire     NaN
4    Charmeleon     fire     NaN
..          ...      ...     ...
796  Celesteela    steel  flying
797     Kartana    grass   steel
798    Guzzlord     dark  dragon
799    Necrozma  psychic     NaN
800    Magearna    steel   fairy

[801 rows x 3 columns]

<br>

### Display all Pokemon whose height is greater than 9 meters with only the name and height attributes.

In [18]:
pokemon[pokemon['height_m'] > 9][['name', 'height_m']]

name  height_m
207     Steelix       9.2
320     Wailord      14.5
796  Celesteela       9.2

#### This code can be broken down and explained as follows:

In [19]:
#This is a boolean indexing operation. 
#It checks each record in the pokemon DataFrame to see if the value in the height_m column is greater than 9. 
#The result is a boolean Series (True/False values) of the same length as the DataFrame.

pokemon['height_m'] > 9

0      False
1      False
2      False
3      False
4      False
       ...  
796     True
797    False
798    False
799    False
800    False
Name: height_m, Length: 801, dtype: bool

In [20]:
#Here, we're using the boolean Series from the previous step to filter the rows of the pokemon DataFrame. 
#Only the rows where the condition (height greater than 9 meters) is True will be retained.

pokemon[pokemon['height_m'] > 9]

abilities  against_bug  against_dark  \
207   ['Rock Head', 'Sturdy', 'Sheer Force']         0.50           1.0   
320  ['Water Veil', 'Oblivious', 'Pressure']         1.00           1.0   
796                          ['Beast Boost']         0.25           1.0   

     against_dragon  against_electric  against_fairy  against_fight  \
207             0.5               0.0            0.5            2.0   
320             1.0               2.0            1.0            1.0   
796             0.5               2.0            0.5            1.0   

     against_fire  against_flying  against_ghost  ...  percentage_male  \
207           2.0             0.5            1.0  ...             50.0   
320           0.5             1.0            1.0  ...             50.0   
796           2.0             0.5            1.0  ...              NaN   

     pokedex_number  sp_attack  sp_defense  speed  type1   type2  weight_kg  \
207             208         55          95     30  steel  ground      400.0   
320             321         90          45     60  water     NaN      398.0   
796             797        107         101     61  steel  flying      999.9   

     generation  is_legendary  
207           2             0  
320           3             0  
796           7             1  

[3 rows x 41 columns]

In [21]:
# After filtering the rows, this part specifies which columns we want to display. 
# In this case, we're interested in the name and height_m columns.
[['name', 'height_m']]

[['name', 'height_m']]

<br>

### Display the index number 207 Pokemon with its name, type1, and type2 attributes.

- `pokemon.loc[...]`: The .loc attribute is used for label-based indexing in pandas.
- 207: This specifies the index label of the row we want to access.
- `['name', 'type1', 'type2']`: This is a list of column names that we want to extract for the specified row.

In [22]:
pokemon.loc[207, ['name', 'type1', 'type2']]


name     Steelix
type1      steel
type2     ground
Name: 207, dtype: object

<br>

### Display the index number 207 Pokemon with attributes of column indices 1,2,3,4, and 0 in order using iloc() 

In [23]:
pokemon.iloc[207,[1,2,3,4,0]]

against_bug                                            0.5
against_dark                                           1.0
against_dragon                                         0.5
against_electric                                       0.0
abilities           ['Rock Head', 'Sturdy', 'Sheer Force']
Name: 207, dtype: object

### Display index number 207 and 208 Pokemon with attributes of column index 1,2,3,4, and 0 in order using iloc() 


In [24]:
pokemon.iloc[[207,208],[1,2,3,4,0]]

against_bug  against_dark  against_dragon  against_electric  \
207          0.5           1.0             0.5               0.0   
208          0.5           0.5             0.0               1.0   

                                  abilities  
207  ['Rock Head', 'Sturdy', 'Sheer Force']  
208   ['Intimidate', 'Run Away', 'Rattled']

<br>

### Display all of the index 207 and 208 Pokemon whose weight_kg is greater than or equal to 400kg with attributes of column index 1,2,3,4, and 0 in order using iloc()


In [25]:
pokemon.iloc[[207,208],[1,2,3,4,0]][pokemon.weight_kg >= 400]

/var/folders/t6/jzgwrdks2vq53w20jlljrk840000gn/T/ipykernel_83888/4126278142.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pokemon.iloc[[207,208],[1,2,3,4,0]][pokemon.weight_kg >= 400]


against_bug  against_dark  against_dragon  against_electric  \
207          0.5           1.0             0.5               0.0   

                                  abilities  
207  ['Rock Head', 'Sturdy', 'Sheer Force']